### ISSUE TITLE
ZWARN information used by mtl ledgers

### ISSUE THREAD

**Issue by ashleyjross** [desihub/desispec#1187]
We need to make sure that the ZWARN bit indicating hardware failures (positioning, fiber, spectrograph, etc.) are set and easily interpretable for updating mtl ledgers. A (very low) tsnr threshold could also be considered as a possible bit (which could be the only one for mtl ledgers to care about, assuming tsnr catches both positioning and spectrograph issues).

This would be very good to have for LSS catalogs for the 1% survey. (It is not good for LSS catalogs if, e.g., failing delta chi2 threshold cuts not due to hardware issues causes re-observation.)

**Comment by sbailey** [comment]
Itemizing the current situation for what bits of ZWARN are (not) set:

Set if FIBERSTATUS != 0, i.e. a known bad fiber or positioner:
```
LITTLE_COVERAGE   = 2**1  #- too little wavelength coverage
NODATA            = 2**9  #- No data for this fiber, e.g. because spectrograph was broken during this exposure (ivar=0 for all pixels)
```

Side effect that occurs when FIBERSTATUS != 0, but can also occur due to other bad/noisy data even if FIBERSTATUS==0:
```
Z_FITLIMIT        = 2**5  #- chi-squared minimum at edge of the redshift fitting range
BAD_MINFIT        = 2**10 #- Bad parabola fit to the chi2 minimum
```

Set when two solutions are within DELTACHI2<9
```
SMALL_DELTA_CHI2  = 2**2  #- chi-squared of best fit is too close to that of second best
```

Currently not set:
```
SKY               = 2**0  #- sky fiber
NEGATIVE_MODEL    = 2**3  #- synthetic spectrum is negative
MANY_OUTLIERS     = 2**4  #- fraction of points more than 5 sigma away from best model is too large (>0.05)
NEGATIVE_EMISSION = 2**6  #- a QSO line exhibits negative emission, triggered only in QSO spectra, if  C_IV, C_III, Mg_II, H_beta, or H_alpha has LINEAREA + 3 * LINEAREA_ERR < 0
UNPLUGGED         = 2**7  #- the fiber was unplugged/broken, so no spectrum obtained
BAD_TARGET        = 2**8  #- catastrophically bad targeting data
```

@ashleyjross @geordie666 for the purposes of MTL re-observation decisions, you could use bit 7 `NODATA` to identify cases where we failed to get a useable spectrum of the target in the first place, and ((bits 2,5, or 10) and not bit 7) for cases where we got a spectrum but were unable fit a confidently good redshift, possibly augmenting or replacing bit 2 (ZWARN) with a different DELTACHI2 cut instead if there is something more appropriate for a particular target class.

These bits are defined by name in `redrock.zwarning.ZWarningMask` [[link](https://github.com/desihub/redrock/blob/master/py/redrock/zwarning.py)]

Commentary on the bits that aren't set:
* `SKY`: we could set this.  See recent desi-data from @stephjuneau regarding this
* `NEGATIVE_MODEL`: I'm hesitant to set this until we've sorted out underlying calibration problems that can cause amp/sky oversubtraction resulting in an offset/negative but otherwise ok spectrum
* `MANY_OUTLIERS`: we could set this, but would need a study of what "many" means
* `NEGATIVE_EMISSION`: we currently penalize chi2 for negative [OII] emission, but don't flip this bit if the selected answer is still negative.  This could be expanded to other lines as well, but would need careful study.
* `UNPLUGGED`: we really should set this bit too when FIBERSTATUS!=0, and probably rename the bit
* `BAD_TARGET`: we could set this if targeting bits `BAD_SKY` or `NO_TARGET` are set (though for `NO_TARGET` we'd need to also make sure that fiberassign is setting that bit when it makes up a location for a positioner that can't reach any targets)

We are not currently propagating TSNR2 into the zbest table, but could.   Summed TSNR2 values are available in the fibermaps of the coadd files in the same directory, so I think this is a (reasonable) bookkeeping convenience but not a blocking factor.  Since it breaks the ability to stack previously generated zbest files, I want to proceed with a bit of caution there.  Let's take that up separately.

**Comment by ashleyjross** [comment]
Stephen, you meant 9 instead of 7 in the text you wrote, correct?

From: Stephen Bailey ***@***.***>
Date: Thursday, April 1, 2021 at 2:36 PM
To: desihub/desispec ***@***.***>
Cc: ashleyjross ***@***.***>, Mention ***@***.***>
Subject: Re: [desihub/desispec] ZWARN information used by mtl ledgers (#1187)

Itemizing the current situation for what bits of ZWARN are (not) set:

Set if FIBERSTATUS != 0, i.e. a known bad fiber or positioner:

LITTLE_COVERAGE   = 2**1  #- too little wavelength coverage

NODATA            = 2**9  #- No data for this fiber, e.g. because spectrograph was broken during this exposure (ivar=0 for all pixels)

Side effect that occurs when FIBERSTATUS != 0, but can also occur due to other bad/noisy data even if FIBERSTATUS==0:

Z_FITLIMIT        = 2**5  #- chi-squared minimum at edge of the redshift fitting range

BAD_MINFIT        = 2**10 #- Bad parabola fit to the chi2 minimum

Set when two solutions are within DELTACHI2<9

SMALL_DELTA_CHI2  = 2**2  #- chi-squared of best fit is too close to that of second best

Currently not set:

SKY               = 2**0  #- sky fiber

NEGATIVE_MODEL    = 2**3  #- synthetic spectrum is negative

MANY_OUTLIERS     = 2**4  #- fraction of points more than 5 sigma away from best model is too large (>0.05)

NEGATIVE_EMISSION = 2**6  #- a QSO line exhibits negative emission, triggered only in QSO spectra, if  C_IV, C_III, Mg_II, H_beta, or H_alpha has LINEAREA + 3 * LINEAREA_ERR < 0

UNPLUGGED         = 2**7  #- the fiber was unplugged/broken, so no spectrum obtained

BAD_TARGET        = 2**8  #- catastrophically bad targeting data

@ashleyjross<https://github.com/ashleyjross> @geordie666<https://github.com/geordie666> for the purposes of MTL re-observation decisions, you could use bit 7 NODATA to identify cases where we failed to get a useable spectrum of the target in the first place, and ((bits 2,5, or 10) and not bit 7) for cases where we got a spectrum but were unable fit a confidently good redshift, possibly augmenting or replacing bit 2 (ZWARN) with a different DELTACHI2 cut instead if there is something more appropriate for a particular target class.

These bits are defined by name in redrock.zwarning.ZWarningMask [link<https://github.com/desihub/redrock/blob/master/py/redrock/zwarning.py>]

Commentary on the bits that aren't set:
·         SKY: we could set this. See recent desi-data from @stephjuneau<https://github.com/stephjuneau> regarding this
·         NEGATIVE_MODEL: I'm hesitant to set this until we've sorted out underlying calibration problems that can cause amp/sky oversubtraction resulting in an offset/negative but otherwise ok spectrum
·         MANY_OUTLIERS: we could set this, but would need a study of what "many" means
·         NEGATIVE_EMISSION: we currently penalize chi2 for negative [OII] emission, but don't flip this bit if the selected answer is still negative. This could be expanded to other lines as well, but would need careful study.
·         UNPLUGGED: we really should set this bit too when FIBERSTATUS!=0, and probably rename the bit
·         BAD_TARGET: we could set this if targeting bits BAD_SKY or NO_TARGET are set (though for NO_TARGET we'd need to also make sure that fiberassign is setting that bit when it makes up a location for a positioner that can't reach any targets)

We are not currently propagating TSNR2 into the zbest table, but could. Summed TSNR2 values are available in the fibermaps of the coadd files in the same directory, so I think this is a (reasonable) bookkeeping convenience but not a blocking factor. Since it breaks the ability to stack previously generated zbest files, I want to proceed with a bit of caution there. Let's take that up separately.

**Comment by sbailey** [comment]
@ashleyjross indeed, yes, I meant bit 9 (NODATA) as the marker for FIBERSTATUS!=0.  Thanks for catching that.  I corrected the comment so that people reading this in the future don't have to apply patches to the post for it to be correct (though the email version already went out...)
